indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java

```diff
@@ -0,0 +1,566 @@
+                                  LOGGER.info("Removing task directory: %s", taskDir);
+                                  FileUtils.deleteDirectory(taskDir);
+                                }
+                              }
+                              catch (Exception e) {
+                                LOGGER.makeAlert(e, "Failed to delete task directory")
+                                      .addData("taskDir", taskDir.toString())
+                                      .addData("task", task.getId())
+                                      .emit();
+                              }
+                            }
+                            catch (Exception e) {
+                              LOGGER.error(e, "Suppressing exception caught while cleaning up task");
+                            }
+                          }
+                        }
+                      }
+                  )
+              )
+      );
+      saveRunningTasks();
+      return tasks.get(task.getId()).getResult();
+    }
+  }
+
+  @Override
+  public void shutdown(String taskid, String reason)
+  {
+    LOGGER.info("Shutdown [%s] because: [%s]", taskid, reason);
+    final ThreadingTaskRunnerWorkItem taskInfo;
+
+    synchronized (tasks) {
+      taskInfo = tasks.get(taskid);
+
+      if (taskInfo == null) {
+        LOGGER.info("Ignoring request to cancel unknown task: %s", taskid);
+        return;
+      }
+
+      if (taskInfo.shutdown) {
+        LOGGER.info(
+            "Task [%s] is already shutting down, ignoring duplicate shutdown request with reason [%s]",
+            taskid,
+            reason
+        );
+      } else {
+        taskInfo.shutdown = true;
+        scheduleTaskShutdown(taskInfo);
+      }
+    }
+  }
+
+  /**
+   * Submits a callable to the control thread pool that attempts a task graceful shutdown,
+   * if shutdown is not already scheduled.
+   *
+   * The shutdown will wait for the configured timeout and then interrupt the thread if the timeout is exceeded.
+   */
+  private ListenableFuture scheduleTaskShutdown(ThreadingTaskRunnerWorkItem taskInfo)
+  {
+    synchronized (tasks) {
+      if (taskInfo.shutdownFuture != null) {
+        return taskInfo.shutdownFuture;
+      }
+
+      ListenableFuture shutdownFuture = controlThreadExecutor.submit(
+          new Callable<Void>()
+          {
+            @Override
+            public Void call()
+            {
+              LOGGER.info("Stopping thread for task: %s", taskInfo.getTaskId());
+              taskInfo.getTask().stopGracefully(taskConfig);
+
+              try {
+                TaskStatus status = taskInfo.getResult().get(
+                    taskConfig.getGracefulShutdownTimeout().toStandardDuration().getMillis(),
+                    TimeUnit.MILLISECONDS
+                );
+
+                if (status == null) {
+                  if (taskInfo.thread != null) {
+                    taskInfo.thread.interrupt();
```

Comment:
is it possible that the interrupt status of the thread is not cleared by this task if it wasn't blocked waiting on something. in that case same thread with interrupt status set would pick a new task ?

Reply:
also I see that you have also updated some of the tasks to capture their own threads on run and interrupt them inside stopGracefully(..) .... one of these appear redundant, no ?

Reply:
For the first point, I added a call to `Thread.interrupted()` at the very end of the outermost finally in the task callable.

For the second point, the interrupt here is meant as a "last resort" if the graceful shutdown didn't complete for some reason within the configured timeout, the interrupt within the tasks should normally end the shutdown before this is called

Reply:
I might be wrong but it appears to me that whole business with the thread is unnecessary and you can achieve same by calling  `ThreadingTaskRunnerWorkItem.shutdownFuture.cance(true)` and you wouldn't have to deal manually with interrupts. That would also remove task from executor if it hasn't started running already.

sidenote: I see that you are trying to simulate force stop of a task which is impossible to do in this context. If task is not `cooperative`, nothing is gonna work unfortunately.

Reply:
also, since we are interrupting the thread from outside, doing same in the task implementations is not achieving anything extra and should probably be removed, that just introduces confusion with no benefit.... or maybe I am still missing something :)